python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll

```diff
@@ -9,3 +9,82 @@ private import semmle.python.dataflow.new.TaintTracking
 private import semmle.python.dataflow.new.RemoteFlowSources
 private import experimental.semmle.python.Concepts
 private import semmle.python.ApiGraphs
+
+/**
+ * Provides models for Python's `re` library.
+ *
+ * See https://docs.python.org/3/library/re.html
+ */
+private module Re {
+  /**
+   * List of `re` methods immediately executing an expression.
+   *
+   * See https://docs.python.org/3/library/re.html#module-contents
+   */
+  private class RegexExecutionMethods extends string {
+    RegexExecutionMethods() {
+      this in ["match", "fullmatch", "search", "split", "findall", "finditer", "sub", "subn"]
+    }
+  }
+
+  /**
+   * A class to find `re` methods immediately executing an expression.
+   *
+   * See `RegexExecutionMethods`
+   */
+  private class DirectRegex extends DataFlow::CallCfgNode, RegexExecution::Range {
+    DataFlow::Node regexNode;
+
+    DirectRegex() {
+      this = API::moduleImport("re").getMember(any(RegexExecutionMethods m)).getACall() and
+      regexNode = this.getArg(0)
+    }
+
+    override DataFlow::Node getRegexNode() { result = regexNode }
+
+    override string getRegexModule() { result = "re" }
+  }
+
+  /**
+   * A class to find `re` methods immediately executing a compiled expression by `re.compile`.
+   *
+   * See `RegexExecutionMethods`
+   *
+   * See https://docs.python.org/3/library/re.html#regular-expression-objects
+   */
+  private class CompiledRegex extends DataFlow::CallCfgNode, RegexExecution::Range {
+    DataFlow::Node regexNode;
+    DataFlow::CallCfgNode regexMethod;
+
+    CompiledRegex() {
+      exists(DataFlow::CallCfgNode patternCall, DataFlow::AttrRead reMethod |
```

Comment:
Since there are multiple flows involved here, I might add a comment with some example syntax to this charred. Say, for the program
```python
   pattern = re.compile(input)
   input.match(s)
```
- `patternCall` refers to `re.compile(input)`
- `regexNode` refers to `input`
- `this` will refer to `input.match(s)`

Reply:
Never done a comment containing code, so feel free to re-review this commit https://github.com/github/codeql/pull/5442/commits/21e01b809fe169f281dbc539e1cc6460dd22edf3 :)

Reply:
I made a comment on your comment, the markdown part is fine :-) Feel free to re-request review when you are ready.

Reply:
I've tried to strictly reference the stuff in scope (https://github.com/github/codeql/pull/5442/commits/213d011a8cfcd3179c7db3922fbe259ddc48df0d). Should I also mention `getRegexModule()` even though it's a hardcoded value? :)

Reply:
You could, since the user will not know that it is hard-coded. But I think the important bits are mentioned now. This class has a slightly complicated characteristic predicate (or charpred), involving flow, so I felt it warranted a bit of explanation.

Reply:
Sounds good 👍. Thanks for your help on the comment 😄, looking forward to the next reviews!